bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/factory/CaddxHandlerFactory.java

```diff
@@ -0,0 +1,263 @@
+import org.openhab.binding.caddx.internal.handler.ThingHandlerKeypad;
+import org.openhab.binding.caddx.internal.handler.ThingHandlerPanel;
+import org.openhab.binding.caddx.internal.handler.ThingHandlerPartition;
+import org.openhab.binding.caddx.internal.handler.ThingHandlerZone;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link CaddxHandlerFactory} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Georgios Moutsos - Initial contribution
+ */
+@Component(configurationPid = "binding.caddx", service = ThingHandlerFactory.class)
+@NonNullByDefault
+public class CaddxHandlerFactory extends BaseThingHandlerFactory {
+
+    private final Logger logger = LoggerFactory.getLogger(CaddxHandlerFactory.class);
+    private final Map<ThingUID, ServiceRegistration<?>> discoveryServiceRegistrations = new HashMap<>();
+    private @NonNullByDefault({}) SerialPortManager portManager;
+
+    @Override
+    public @Nullable Thing createThing(ThingTypeUID thingTypeUID, Configuration configuration,
+            @Nullable ThingUID thingUID, @Nullable ThingUID bridgeUID) {
+        logger.trace("createThing(): thingTypeUID- {}, configuration - {}", thingTypeUID, configuration);
+
+        if (CaddxBindingConstants.CADDXBRIDGE_THING_TYPE.equals(thingTypeUID)) {
+            ThingUID caddxBridgeUID = getCaddxBridgeThingUID(thingTypeUID, thingUID, configuration);
+            return super.createThing(thingTypeUID, configuration, caddxBridgeUID, null);
+        } else if (CaddxBindingConstants.PANEL_THING_TYPE.equals(thingTypeUID)) {
+            ThingUID panelThingUID = getCaddxPanelUID(thingTypeUID, thingUID, configuration, bridgeUID);
+            return super.createThing(thingTypeUID, configuration, panelThingUID, bridgeUID);
+        } else if (CaddxBindingConstants.PARTITION_THING_TYPE.equals(thingTypeUID)) {
+            ThingUID partitionThingUID = getCaddxPartitionUID(thingTypeUID, thingUID, configuration, bridgeUID);
+            return super.createThing(thingTypeUID, configuration, partitionThingUID, bridgeUID);
+        } else if (CaddxBindingConstants.ZONE_THING_TYPE.equals(thingTypeUID)) {
+            ThingUID zoneThingUID = getCaddxZoneUID(thingTypeUID, thingUID, configuration, bridgeUID);
+            return super.createThing(thingTypeUID, configuration, zoneThingUID, bridgeUID);
+        } else if (CaddxBindingConstants.KEYPAD_THING_TYPE.equals(thingTypeUID)) {
+            ThingUID keypadThingUID = getCaddxKeypadUID(thingTypeUID, thingUID, configuration, bridgeUID);
+            return super.createThing(thingTypeUID, configuration, keypadThingUID, bridgeUID);
+        }
+
+        throw new IllegalArgumentException(
+                "createThing(): The thing type " + thingTypeUID + " is not supported by the Caddx binding.");
```

Comment:
You shouldn't throw an exception here, instead you should return null. You can log if this ever happens.

Reply:
Done